Plugins/NativeEngine/Source/NativeDataStream.h

```diff
@@ -0,0 +1,223 @@
+            }
+
+            gsl::span<int32_t> ReadInt32Array()
+            {
+                Validate<ValidationType::Int32Array>(*this);
+                return ReadArray<int32_t>();
+            }
+
+            gsl::span<float> ReadFloat32Array()
+            {
+                Validate<ValidationType::Float32Array>(*this);
+                return ReadArray<float>();
+            }
+
+            template<typename T>
+            T ReadNativeData()
+            {
+                Validate<ValidationType::NativeData>(*this);
+                const auto data = ReadUint32Array();
+                return *reinterpret_cast<T*>(data.data());
+            }
+
+            template<typename T, class = typename std::enable_if<!std::is_pointer<T>::value>::type>
+            auto ReadPointer()
+            {
+                return ReadNativeData<typename std::conditional<std::is_member_pointer<T>::value, T, T*>::type>();
+            }
+
+        private:
+            gsl::span<uint8_t> m_bytes{};
+            size_t m_position{0};
+            const gsl::final_action<std::function<void()>> m_scopeGuard;
+
+            friend class NativeDataStream;
+
+            template<typename CallableT>
+            Reader(gsl::span<uint8_t> bytes, CallableT&& callable)
+                : m_bytes{bytes}
+                , m_scopeGuard{std::forward<CallableT>(callable)}
+            {
+            }
+
+            template<typename T>
+            T Read()
+            {
+                T t{*reinterpret_cast<T*>(m_bytes.data() + m_position)};
+                m_position += sizeof(T);
+                return t;
+            }
+
+            template<typename T>
+            gsl::span<T> ReadArray()
+            {
+                // Arrays are byte aligned
+                ByteAlign();
+                // The first 32 bit number is a size
+                uint32_t size = Read<uint32_t>();
+
+                auto span = gsl::make_span<T>(reinterpret_cast<T*>(m_bytes.data() + m_position), size);
+                m_position += size * sizeof(T);
+                return span;
+            }
+
+            void ByteAlign()
+            {
+                // TODO: Put back byte alignment when we're doing things more efficiently in TS.
+                size_t remainder = m_position % 4 > 0;
+                if (remainder > 0)
+                {
+                    m_position += (4 - remainder);
+                }
+            }
+        };
+
+        static void Initialize(Napi::Env env)
+        {
+            Napi::HandleScope scope{env};
+            if constexpr (VALIDATION_ENABLED)
+            {
+                Napi::Function func = DefineClass(
+                    env,
+                    JS_CLASS_NAME,
+                    {
+                        InstanceMethod("writeBytes", &NativeDataStream::WriteBytes),
+
+                        StaticValue("VALIDATION_UINT_8", Napi::Number::From(env, static_cast<int>(ValidationType::Uint8))),
```

Comment:
Either this should be the following or the enum should derive from `int` to be consistent. Napi should accept either.
```suggestion
                        StaticValue("VALIDATION_UINT_8", Napi::Number::From(env, static_cast<uint32_t>(ValidationType::Uint8))),
```